core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads_wrapper.c

```diff
@@ -123,14 +123,12 @@ thread_spawn_wrapper(wasm_exec_env_t exec_env, uint32 start_arg)
     thread_start_arg->arg = start_arg;
     thread_start_arg->start_func = start_func;
 
-    os_mutex_lock(&exec_env->wait_lock);
     ret = wasm_cluster_create_thread(exec_env, new_module_inst, false,
                                      thread_start, thread_start_arg);
     if (ret != 0) {
         LOG_ERROR("Failed to spawn a new thread");
         goto thread_spawn_fail;
     }
-    os_mutex_unlock(&exec_env->wait_lock);
```

Comment:
Is there data race here?
Had better not remove the lock, instead, add a cond_wait like pthread_create_wrapper in lib_pthread_wrapper.c:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/38c67b3f4815a628ce0f9bc22186bface4678089/core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c#L665-L668

And in the `thread_start`, add cond_signal like pthread_start_routine:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/38c67b3f4815a628ce0f9bc22186bface4678089/core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c#L509

Reply:
The reason why I removed it is that, after getting the `exec_env` lock there, the `cluster` lock gets acquired https://github.com/bytecodealliance/wasm-micro-runtime/blob/38c67b3f4815a628ce0f9bc22186bface4678089/core/iwasm/libraries/thread-mgr/thread_manager.c#L623
but then an attempt to get the `exec_env` lock is made again at https://github.com/eloparco/wasm-micro-runtime/blob/5550b5647037a7cc9674dda03413a95e3c06cba3/core/iwasm/libraries/thread-mgr/thread_manager.c#L946 generating a warning in the sanitizer for potential deadlock.

Let me see how it can be fixed.

Reply:
It seems that the thread isn't actually started, but it is already added to the cluster's exec_env list, and another thread (seems main thread) tries to terminate it? Maybe after adding the similar logic like pthread_create_wrapper, the issue can be resolved since the main thread will wait unit the child thread actually started.
Here try adding cond_wait:
```C
    os_mutex_lock(&exec_env->wait_lock);
    ret = wasm_cluster_create_thread(exec_env, new_module_inst, false,
                                     thread_start, thread_start_arg);
    ...
    os_cond_wait(&exec_env->wait_cond, &exec_env->wait_lock);
    os_mutex_unlock(&exec_env->wait_lock);
```
And in thread_start:
```
    os_mutex_lock(&parent_exec_env->wait_lock);
    ...
    os_cond_signal(&parent_exec_env->wait_cond, &parent_exec_env->wait_lock);
    os_mutex_unlock(&parent_exec_env->wait_lock);
```

Reply:
It actually only happens when it's the spawn thread that calls `proc_exit`.
https://github.com/bytecodealliance/wasm-micro-runtime/blob/38c67b3f4815a628ce0f9bc22186bface4678089/core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c#L665-L668
I see that for pthreads that lock is used to wait for the update of `ThreadInfoNode`.
In `lib_wasi_threads_wrapper.c` we don't use `ThreadInfoNode` so I don't think the lock there is even needed.

Reply:
OK, got it, thanks.

Reply:
Great, once this one is merged, we can merge #1985 too